doc/formatter.md

```diff
@@ -67,6 +67,79 @@ subproblem: _Given a box width constraint, what are the best ways to format a
 region of text?_ Leaf partitions of the [TokenPartitionTree] represents a unit
 of work that is currently handled by the line-wrapping optimization phase.
 
+### Tree unwrapping
+
+The implemention of [TreeUnwrapper] is split into
+[language-agnostic library functions](https://cs.opensource.google/verible/verible/+/master:common/formatting/tree_unwrapper.cc)
+and [Verilog-specific](https://cs.opensource.google/verible/verible/+/master:verilog/formatting/formatter.cc) code.
+
+Simplified class inheritance diagram is shown below:
+![Inheritance diagram](./formatter_simplified_class_diagram.png)
+
+[TreeUnwrapper] class exploits
+the [Visitor](https://github.com/google/verible/blob/ff1c692695ca723275f24397572cb1d1eedb6779/common/text/visitors.h#L44)
+pattern to recursively process the input CST tree:
+```cpp
+// SymbolVisitor is an abstract visitor class from which visitors can be derived
+// SymbolVisitor only visits a single Symbol, and does not recurse on nodes
+//
+// Usage:
+//  SymbolPtr tree = ...
+//  SymbolVisitorSubclass visitor;
+//  tree->Accept(visitor);
+//
+class SymbolVisitor {
+ public:
+  virtual ~SymbolVisitor() {}
+  virtual void Visit(const SyntaxTreeLeaf& leaf) = 0;
+  virtual void Visit(const SyntaxTreeNode& node) = 0;
+};
+```
+
+It builds the [TokenPartitionTree] (i.e. `VectorTree<UnwrappedLine>`) with a set of helper methods to simplify the process:
+* [`TraverseChildren()`](https://github.com/google/verible/blob/ff1c692695ca723275f24397572cb1d1eedb6779/common/formatting/tree_unwrapper.cc#L278) - only traverses the CST
+* [`VisitIndentedSection()`](https://github.com/google/verible/blob/ff1c692695ca723275f24397572cb1d1eedb6779/common/formatting/tree_unwrapper.cc#L324) - creates indented section and traverses the CST
+* `VisitNewUnwrappedLine()` - starts new line (in current indented section) and traverses the CST
+
+The following two code blocks show examples of such:
+```cpp
+void TreeUnwrapper::TraverseChildren(const verible::SyntaxTreeNode& node) {
+  // Can't just use TreeContextVisitor::Visit(node) because we need to
+  // call a visit hook between children.
+  const verible::SyntaxTreeContext::AutoPop p(&current_context_, &node);
+  InterChildNodeHook(node);
+  for (const auto& child : node.children()) {
+    if (child) {
+      child->Accept(this);
+      InterChildNodeHook(node);
+    }
+  }
+}
+```
+
+```cpp
+void TreeUnwrapper::VisitIndentedSection(const SyntaxTreeNode& node,
+                                         int indentation_delta,
+                                         PartitionPolicyEnum partitioning) {
+  const auto last_ftoken_iter =
+      VisitIndentedChildren(node, indentation_delta, partitioning);
+  // TODO(fangism): do we ever need to remove trailing empty partitions here?
```

Comment:
TODO(me)? Well that's embarrassing!  Probably not a good idea to name authors in documentation, even well-intended ones.  :)